### ISSUE TITLE
Use profdata for collecting code coverage for clang

### ISSUE THREAD

**Issue by iirina** [bazelbuild/bazel#5881]
Currently C++ coverage is 2x slower than just running the corresponding tests. The goal is to have similar times for running bazel test and bazel coverage on the same targets. When `clang` is used `profdata` is the fastest and reliable way for computing code coverage results.

**Comment by m3rcuriel** [comment]
@iirina I'm struggling to understand the status of this (and clang coverage in general). I have `--experimental_use_llvm_covmap` generating the `coverage.dat` files (aka the `.profdata`) files. What is the next step in resolving this issue? What functionality am I still missing?

Thanks for your work on coverage!

**Comment by iirina** [comment]
The ideal state that we want to be in is for Bazel to have a standardized coverage output, which is `lcov`. The main thing left here is to have a `.profdata` -> `lcov` converter and move the flag out of experimental (prototype -> proper feature, e.g. declare profdata tools).

**Comment by m3rcuriel** [comment]
Awesome! Can't wait to have standardized coverage, but I'm also perfectly happy about the experimental profdata -- thanks!

**Comment by cmoore3** [comment]
> The ideal state that we want to be in is for Bazel to have a standardized coverage output, which is `lcov`. The main thing left here is to have a `.profdata` -> `lcov` converter and move the flag out of experimental (prototype -> proper feature, e.g. declare profdata tools).

Does `llvm-cov export -format=lcov -instr-profile <profdata file> <relevant object files>` work for that?

**Comment by iirina** [comment]
> Does llvm-cov export -format=lcov -instr-profile <profdata file> <relevant object files> work for that?

Yes it's exactly what we need. The only problem is that `-format=lcov` was added recently and older versions of `llvm-cov` don't support it.

> We were using BAZEL_CC_COVERAGE_TOOL=LCOV and supplying an LCOV_MERGER that wrote the coverage.dat files in lcov format for Bazel 0.25.0. In 0.26, that appears to have been removed...

Yes, that was removed as part of #7759. Doesn't the default gcov coverage work out of the box for you?

**Comment by cmoore3** [comment]
> Yes, that was removed as part of #7759. Doesn't the default gcov coverage work out of the box for you?

Great question!  I just tried it myself and found the same results as a colleague had told me.

Here are the info. messages that I get (using Bazel 0.28.0 with default coverage settings and a clang 8.0.0 crosstool):
```
$ bazel coverage //hello_world/cc:hello_world_test
...
//hello_world/cc:hello_world_test                                        PASSED in 0.7s
...
$ ls -ltr bazel-testlogs/hello_world/cc/hello_world_test/
total 52
drwxrwxr-x 3 cmoore3 abcdef  4096 Jul 11 11:53 ..
-r-xr-xr-x 1 cmoore3 abcdef   167 Jul 11 11:53 baseline_coverage.dat
drwxrwxr-x 2 cmoore3 abcdef  4096 Jul 11 11:53 test.raw_splitlogs
drwxrwxr-x 2 cmoore3 abcdef  4096 Jul 11 11:53 test.outputs_manifest
drwxrwxr-x 2 cmoore3 abcdef  4096 Jul 11 11:53 test.outputs
-rw-rw-r-- 1 cmoore3 abcdef   607 Jul 11 11:53 test.xml
-r-xr-xr-x 1 cmoore3 abcdef     0 Jul 11 11:53 coverage.dat
-r-xr-xr-x 1 cmoore3 abcdef 19779 Jul 11 11:53 test.log
-r-xr-xr-x 1 cmoore3 abcdef   188 Jul 11 11:53 test.cache_status
drwxrwxr-x 5 cmoore3 abcdef  4096 Jul 11 11:53 .
$ cat bazel-testlogs/hello_world/cc/hello_world_test/test.log
...
Jul 11, 2019 3:45:51 PM com.google.devtools.coverageoutputgenerator.Main getTracefiles
INFO: No lcov file found.
Jul 11, 2019 3:45:51 PM com.google.devtools.coverageoutputgenerator.Main getGcovInfoFiles
INFO: No gcov info file found.
Jul 11, 2019 3:45:51 PM com.google.devtools.coverageoutputgenerator.Main getProfdataFileOrNull
INFO: No .profdata file found.
Jul 11, 2019 3:45:51 PM com.google.devtools.coverageoutputgenerator.Main main
WARNING: There was no coverage found.
```
Note that `coverage.dat` is empty and there are not any `.gcda` files anywhere.  There are `.gcno` files that were produced, though.

**Comment by mhines01** [comment]
I am seeing the same issue with:

mhines@box:~/src/ebplane$ bazel coverage ... --experimental_use_llvm_covmap 
Starting local Bazel server and connecting to it...
INFO: Using default value for --instrumentation_filter: "//daemon[/:]".
INFO: Override the above default with --instrumentation_filter
INFO: Analyzed 3 targets (26 packages loaded, 957 targets configured).
INFO: Found 2 targets and 1 test target...
INFO: Deleting stale sandbox base /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox
INFO: Elapsed time: 3.100s, Critical Path: 0.26s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
PASSED: //daemon:main_test (see /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/execroot/__main__/bazel-out/k8-fastbuild/testlogs/daemon/main_test/test.log)
INFO: From Testing //daemon:main_test
==================== Test output for //daemon:main_test:
+ [[ -z bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files ]]
+ export ROOT=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__
+ ROOT=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__
+ [[ bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files != /* ]]
+ export COVERAGE_MANIFEST=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files
+ COVERAGE_MANIFEST=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files
+ export COVERAGE_DIR=_coverage/daemon/main_test/test
+ COVERAGE_DIR=_coverage/daemon/main_test/test
+ [[ _coverage/daemon/main_test/test == /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__* ]]
+ COVERAGE_DIR=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test
+ mkdir -p /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test
+ COVERAGE_OUTPUT_FILE=bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat
+ [[ bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat == /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__* ]]
+ COVERAGE_OUTPUT_FILE=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat
+ export JAVA_COVERAGE_FILE=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/jvcov.dat
+ JAVA_COVERAGE_FILE=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/jvcov.dat
+ export COVERAGE=1
+ COVERAGE=1
+ export BULK_COVERAGE_RUN=1
+ BULK_COVERAGE_RUN=1
+ for name in "$LCOV_MERGER"
+ [[ ! -e bazel-out/k8-fastbuild/bin/external/bazel_tools/tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main ]]
+ export GCOV_PREFIX_STRIP=3
+ GCOV_PREFIX_STRIP=3
+ export GCOV_PREFIX=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test
+ GCOV_PREFIX=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test
+ export LLVM_PROFILE_FILE=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/%h-%p-%m.profraw
+ LLVM_PROFILE_FILE=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/%h-%p-%m.profraw
+ [[ ! -z '' ]]
+ cd /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.runfiles/__main__
+ /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.runfiles/__main__/daemon/main_test
Running main() from gmock_main.cc
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from MainTest
[ RUN      ] MainTest.TestExample
[       OK ] MainTest.TestExample (0 ms)
[----------] 1 test from MainTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (0 ms total)
[  PASSED  ] 1 test.
+ TEST_STATUS=0
+ touch /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat
+ [[ 0 -ne 0 ]]
+ cd /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__
+ [[ -n external/bazel_tools/tools/test/collect_cc_coverage.sh ]]
+ eval external/bazel_tools/tools/test/collect_cc_coverage.sh
++ external/bazel_tools/tools/test/collect_cc_coverage.sh
+ main
+ init_gcov
+ GCOV=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/gcov
+ ln -s /usr/bin/gcov /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/gcov
+ uses_llvm
+ stat '/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/*.profraw'
+ return 1
+ case "$BAZEL_CC_COVERAGE_TOOL" in
+ gcov_coverage /home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/_cc_coverage.gcov
+ local output_file=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/_cc_coverage.gcov
+ shift
+ local gcov_log=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test/_cc_coverage.gcov.gcov.log
+ read -r line
+ LCOV_MERGER_CMD='bazel-out/k8-fastbuild/bin/external/bazel_tools/tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main --coverage_dir=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test   --output_file=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat   --filter_sources=/usr/bin/.+   --filter_sources=/usr/lib/.+   --filter_sources=/usr/include.+   --filter_sources=.*external/.+   --source_file_manifest=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files'
+ [[ -n '' ]]
+ [[ -n '' ]]
+ JAVA_RUNFILES=
+ exec bazel-out/k8-fastbuild/bin/external/bazel_tools/tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/Main --coverage_dir=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/_coverage/daemon/main_test/test --output_file=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/testlogs/daemon/main_test/coverage.dat --filter_sources=/usr/bin/.+ --filter_sources=/usr/lib/.+ --filter_sources=/usr/include.+ '--filter_sources=.*external/.+' --source_file_manifest=/home/mhines/.cache/bazel/_bazel_mhines/1d3eaee6c73d908b7b25eba523cf0ef6/sandbox/linux-sandbox/29/execroot/__main__/bazel-out/k8-fastbuild/bin/daemon/main_test.instrumented_files
Aug 09, 2019 3:43:27 AM com.google.devtools.coverageoutputgenerator.Main getTracefiles
INFO: No lcov file found.
Aug 09, 2019 3:43:27 AM com.google.devtools.coverageoutputgenerator.Main getGcovInfoFiles
INFO: No gcov info file found.
Aug 09, 2019 3:43:27 AM com.google.devtools.coverageoutputgenerator.Main getProfdataFileOrNull
INFO: No .profdata file found.
Aug 09, 2019 3:43:27 AM com.google.devtools.coverageoutputgenerator.Main main
WARNING: There was no coverage found.
================================================================================
//daemon:main_test                                              (cached) PASSED in 0.2s

Executed 0 out of 1 test: 1 test passes.
INFO: Build completed successfully, 1 total action

mhines@box:~/src/ebplane$ bazel version
Build label: 0.28.1
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Jul 19 15:19:51 2019 (1563549591)
Build timestamp: 1563549591
Build timestamp as int: 1563549591

**Comment by nnooney** [comment]
I ran into this issue as well -- the .gcno files are generated, but there are no .gcda files. It turns out that during linking of the object files, the path to the .gcda file is hard-coded into the resulting object. Inspecting this string via `strings hello_world_test | grep gcda` shows the path of the sandbox which was used to compile the intermediate object. The .gcda file does exist at that path after running the test object.

However, the collect_cc_coverage.sh script in Bazel Tools expects that .gcda file to be generated at the same path as the .gcno file. I'm not experienced enough with Bazel to understand/implement the changes necessary for this to work.

**Comment by emusand** [comment]
> Yes it's exactly what we need. The only problem is that -format=lcov was added recently and older versions of llvm-cov don't support it.

My organization is using a newer clang version, which supports `llvm-cov export --format=lcov`. Is there any way to get `bazel coverage --experimental_use_llvm_covmap` to use that command to produce lcov reports?

Is the plan to use `llvm-cov export --format=lcov ...` if supported, and fall back on a custom `.profdata` -> `lcov` converter if not?

**Comment by garethellis0** [comment]
My teams workaround for now is to manually find the `coverage.dat` files, merge them, and convert to `lcov`. Obviously there are a lot of problems with doing it this way, but it works for us for now.

```
#
# This script finds `.dat` files, assumes they are `profdata` files, and
# converts them to a `.lcov` file that CodeCov can understand. 
# 
# This script exists because clang outputs a different format for code coverage
# data then gcc does, and bazel has not yet unified it's coverage output.
# Once https://github.com/bazelbuild/bazel/issues/9406 is resolved we should be
# able to remove this script
#

MERGED_PROFDATA_FILE_NAME="merged_coverage.dat"
MERGED_LCOV_FILE_NAME="$MERGED_PROFDATA_FILE_NAME.lcov"
PROFDATA_FILE_LIST_NAME="all_coverage_files.txt"

# All the folders we'll search for object files in
declare -a OBJECT_FILES_FOLDERS=(
    "bazel-out/k8-fastbuild/bin/software"
    "bazel-out/k8-fastbuild/bin/firmware_new"
    "bazel-out/k8-fastbuild/bin/firmware"
)

# Some object files seem to make `llvm-cov` fail with the following error:
# "error: Could not load coverage information"
# So we ignore them.
# At the time of writing, the known bad files are:
#   bazel-out/k8-fastbuild/bin/software/gui/widgets/_objs/parameters/moc_parameters.pic.o 
#   bazel-out/k8-fastbuild/bin/software/gui/widgets/_objs/robot_status/moc_robot_status.pic.o 
#   bazel-out/k8-fastbuild/bin/software/gui/widgets/_objs/world_view/moc_world_view.pic.o 
#   bazel-out/k8-fastbuild/bin/software/gui/widgets/_objs/ai_control/moc_ai_control.pic.o 
# Please refer to this post for how to construct a prune argument. It's not
# as simple as you think. https://stackoverflow.com/a/4210072
BLACKLIST_ARG="-path bazel-out/k8-fastbuild/bin/software/gui/widgets -prune -o"

# Find all object files
declare -a OBJECT_FILES=()
for folder in "${OBJECT_FILES_FOLDERS[@]}"; do
    for file in $(find -L $folder $BLACKLIST_ARG -name '*.o' -type f -print); do 
        OBJECT_FILES+=( "$file" )
    done
done

# Create a list of arguments like "-o my_file1.o -o my_other_file.o ..." that 
# we can pass into `llvm-cov`
OBJECT_FILES_ARG=""
for object_file in "${OBJECT_FILES[@]}"; do
    OBJECT_FILES_ARG="$OBJECT_FILES_ARG -object $object_file"
done

# Find all profdata files (we assume bazel was configured to output profdata
# to all the ".dat" files, there's no easy way to check that it's not just an
# lcov file already), and put the found paths into a file
# Create a file with the names of all the coverage files we want to merge
rm -rf $PROFDATA_FILE_LIST_NAME
for profdata_file_path in $(find -L bazel-testlogs -iname "coverage.dat"); do
    echo "$profdata_file_path" >> $PROFDATA_FILE_LIST_NAME
done

# Merge all the profdata files into one big one
rm -rf $MERGED_PROFDATA_FILE_NAME
./bazel-src/external/llvm_clang/bin/llvm-profdata merge -output=$MERGED_PROFDATA_FILE_NAME -input-files=$PROFDATA_FILE_LIST_NAME

# Convert the merged profdata file into a single lcov file
rm -rf $MERGED_LCOV_FILE_NAME
./bazel-src/external/llvm_clang/bin/llvm-cov export -format=lcov -instr-profile="$MERGED_PROFDATA_FILE_NAME" $OBJECT_FILES_ARG > "$MERGED_PROFDATA_FILE_NAME.lcov"

# Clean up intermediate files, otherwise CodeCov will pick these up as well
rm -rf $PROFDATA_FILE_LIST_NAME
rm -rf $MERGED_PROFDATA_FILE_NAME
```

**Comment by MaskRay** [comment]
I have made a bunch of improvement to clang HEAD. With the next release of clang (11.0.0), `clang --coverage` will emit gcov 4.8 format .gcno/.gcda (which is compatible with gcov 4.8~7)

* `-Xclang -coverage-version='304*'` chooses 3.4 (compatible with gcov 3.4~4.6)
* `-Xclang -coverage-version='407*'` chooses 4.7
* default: 4.8 (compatible with gcov 4.8~7)
* `-Xclang -coverage-version='800*'` chooses 8.0
* `-Xclang -coverage-version='900*'` chooses 9.0 (compatible with gcov 9~10)

**Comment by c-mita** [comment]
I believe this is resolved by 21b5eb62. We may need to watch out for the aforementioned clang coverage change however.
